### ISSUE TITLE
DatePicker: Add minTime and maxTime attributes

### ISSUE THREAD

**Issue by tuerker** [primefaces/primefaces#5551]
Currently, the p:calendar component provides the attributes minHour and minMinute as well as maxHour and maxMinute, but these attributes cannot really be used to restrict a valid time span from which a user might select the time.

Assume, I want to restrict the time the user can choose to any time in the time interval 09:00-17:00.
If I use minHour="9", minMinute="0", maxHour="17" and maxMinute="0", then the time span is restricted to 09:00-17:00 but I cannot choose any minutes else than 0. So, 9:30 or 10:45 is not possible anymore.

Feature Request: So, I would like to ask to add the attributes minTime and maxTime to p:calendar component to allow to restrict a valid time span to a given minimum and maximum time which are instances of class LocalTime.

## 1) Environment
- PrimeFaces version: 8.0

## 2) Expected behavior

If minTime is set, the user cannot select a time before minTime.

If maxTime is set, the user cannot select a time after maxTime.

If both are set, enforce that minTime is before maxTime

## 3) Actual behavior

..

## 4) Steps to reproduce

.. 

## 5) Sample XHTML

..

## 6) Sample bean

..

**Comment by christophs78** [comment]
@tuerker : You may use the DatePicker-component. DatePicker is the succesor of Calendar and already has this feature. (see https://www.primefaces.org/showcase/ui/input/datepicker/datePickerJava8.xhtml)
Just wait for the PrimeFaces 8.0-release. (DatePicker in PrimeFaces 7.0 is buggy)

**Comment by tuerker** [comment]
Dear christoph, thanks for the hint. We will return back to the datePicker (which we replaced by the Calendar because the datepicker was buggy in PF7).

Nevertheless, the datepicker also does not support the requested feature. As in the calendar component, there are only mindate and maxdate attributes but i am requiring a datetime in a given mindate and maxdate interval and ontop of that only times between a certain time interval should be allowed.

Suppose i am implementing an instrument reservation where i want to state that in the year 2020 the reservation times are between 09:00 and 17:00 while in 2021 (or in a certain month or whatever period) the allowed times should 08:00-19:00.

The attributes minDate and maxDate alone are not sufficient to realize this. Additional minTime and maxTime attributes are needed. If I have overlooked something, i would really appreciate to hear how this scenario could be implemented with the current available features.

Many thanks in advance. Best regards, Can.

**Comment by christophs78** [comment]
mindate- and maxdate-DatePicker-attributes are a bit missleading. You can bind them to java.time.LocalTime. (see https://primefaces.github.io/primefaces/8_0/#/components/datepicker)
So you may have one datePicker with a value-binding to java.time.LocalDate and another datePicker (with timeOnly=true and minDate=someLocalTime and maxDate=someLocalTime) with a value-binding to java.time.LocalTime.
Could that work for you?

**Comment by tuerker** [comment]
No, i want to have a single datepicker where on the upper part i select the date and in the lower part i select the time. the add-on that is needed is the selectable time must come from a given interval set by mintime and maxtime.

**Comment by tuerker** [comment]
This is similar to the minHour and maxHour of p:calendar but goes beyond by providing the manTim and maxTime as LocalTime values.

**Comment by christophs78** [comment]
I understand your usecase. And i also think it´s really useful.
Just try to think what we can/should do. :-)

1) PF-Calendar is based on jQueryUI. And that´s problematic because jQueryUI is almost dead since 2017. This is why PF-DatePicker was introduced. (There is only minimal life support to jQueryUI to keep it working with newer jQuery-versions.)
But https://trentrichardson.com/examples/timepicker/ which is the base of PF-Calendar already supports minTime and maxTime. We just would need to wrap it in PrimeFaces. (--> "quick win")

2) To be future-proof we would need to add this to PF-DatePicker. But that´s a thing, which is not so simple. (To some degree because we should not break any existing working UseCases.)

@tuerker : Can you provide a PR for 2? Or is it an option to pay ("sponsor") PrimeTek for adding this feature?

@tandraschko: What do you think about adding features to PF-Calendar? This sounds easy and is something i may do in my spare time. Just not future-proof.

**Comment by tandraschko** [comment]
@christophs78 in general i would not further enhance p:calendar with own code, we should try to make datePicker better

**Comment by tuerker** [comment]
i think that tandraschko's proposal should be the way to go. 

in the meanwhile I will solve my problem with a validator which is anyhow needed for realizing other constraints, too. 

many thanks for your efforts.

**Comment by tuerker** [comment]
i just changed the title of this issue since it make more sense to enhance the datePicker instead of the old calendar component.
